### PR TITLE
DMDA

### PR DESCRIPTION
**Pull request by mrogowski**

### CODE REVIEW

**Comment by mrogowski** [comment]
Ready for review @dalcinl @stefanozampini @nocollier.

File: src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx [hunk 1]
```diff
@@ -157,7 +305,16 @@ cdef class DMDA(DM):
             CHKERR( DMSetDimension(self.dm, gdim) )
         CHKERR( DMDASetSizes(self.dm, M, N, P) )
 
-    def getSizes(self):
+    def getSizes(self) -> tuple[int, ...]:
+        """Return the global dimension in one/two/three directions.
```
Comment (stefanozampini) [inline src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx:309]:
```suggestion
        """Return the number of grid points in each direction.
```

File: src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx [hunk 2]
```diff
@@ -171,7 +328,21 @@ cdef class DMDA(DM):
                             NULL) )
         return toDims(dim, M, N, P)
 
-    def setProcSizes(self, proc_sizes):
+    def setProcSizes(self, proc_sizes: tuple[int, ...]) -> None:
+        """Set the number of processes in each dimension.
```
Comment (stefanozampini) [inline src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx:332]: It seems you use `dimensions` and `directions` almost interchangeably. I'm ok with both, but we have to be consistent and use only one.

File: src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx [hunk 3]
```diff
@@ -219,11 +426,34 @@ cdef class DMDA(DM):
                           NULL) )
         return toDims(dim, btx, bty, btz)
 
-    def setStencilType(self, stencil_type):
+    def setStencilType(self, stencil_type: StencilType) -> None:
+        """Set the type of the communication stencil.
```
Comment (stefanozampini) [inline src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx:430]:
```suggestion
        """Set the stencil type.
```

File: src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx [hunk 4]
```diff
@@ -234,11 +464,34 @@ cdef class DMDA(DM):
                           &stype) )
         return stype
 
-    def setStencilWidth(self, stencil_width):
+    def setStencilWidth(self, stencil_width: int) -> None:
+        """Set the width of the communication stencil.
```
Comment (stefanozampini) [inline src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx:468]:
```suggestion
        """Set the stencil width.
```

File: src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx [hunk 5]
```diff
@@ -269,7 +552,19 @@ cdef class DMDA(DM):
 
     #
 
-    def getRanges(self):
+    # TODO: should we clarify how is getRanges different than getGhostRanges and getCorners?
```
Comment (stefanozampini) [inline src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx:555]: I don't think so. Just refer to the dmda manual section at the top of the class and refer to the specifi C api in each function

Comment (stefanozampini) [reply src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx:555]: maybe list all possible python getters in the see also

File: src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx [hunk 6]
```diff
@@ -269,7 +552,19 @@ cdef class DMDA(DM):
...
+    def getRanges(self) -> tuple[tuple[int, int], ...]:
+        """Return the ranges of the local region in each dimension.
```
Comment (stefanozampini) [inline src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx:557]:
```suggestion
        """Return the ranges of the owned local region in each dimension.
```

File: src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx [hunk 7]
```diff
@@ -302,7 +618,26 @@ cdef class DMDA(DM):
         CHKERR( DMDAGetOwnershipRanges(self.dm, &lx, &ly, &lz) )
         return toOwnershipRanges(dim, m, n, p, lx, ly, lz)
 
-    def getCorners(self):
+    def getCorners(self) -> tuple[tuple[int, ...], tuple[int, ...]]:
+        """Return lower left corner and size of local region in each dimension.
```
Comment (stefanozampini) [inline src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx:622]:
```suggestion
        """Return the lower left corner and the size of the owned local region in each dimension.
```

File: src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx [hunk 8]
```diff
@@ -311,7 +646,26 @@ cdef class DMDA(DM):
         return ((toInt(x), toInt(y), toInt(z))[:<Py_ssize_t>dim],
                 (toInt(m), toInt(n), toInt(p))[:<Py_ssize_t>dim])
 
-    def getGhostCorners(self):
+    def getGhostCorners(self) -> tuple[()] | tuple[tuple[int], tuple[int]] | tuple[tuple[int, int], tuple[int, int]] | tuple[tuple[int, int, int], tuple[int, int, int]]:
+        """Return lower left corner and size of local region in each dimension.
```
Comment (stefanozampini) [inline src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx:650]:
```suggestion
        """Return the lower left corner and the size of the local region in each dimension, including ghost points.
```

File: src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx [hunk 9]
```diff
@@ -311,7 +646,26 @@ cdef class DMDA(DM):
...
+
+        Returns the global (x,y,z) indices of the lower left corner (first
+        tuple) and size of the local region (second tuple).
+
+        Excluding ghost points.
+
+        The corner information is independent of the number of degrees of
```
Comment (stefanozampini) [inline src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx:657]: No need?

File: src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx [hunk 10]
```diff
@@ -322,29 +676,108 @@ cdef class DMDA(DM):
 
     #
 
-    def setFieldName(self, field, name):
+    def setFieldName(self, field: int, name: str) -> None:
+        """Set the name of individual field components.
+
+        Logically collective; ``name`` must contain a common value.
```
Comment (stefanozampini) [inline src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx:682]:
```suggestion
        Logically collective.
```

File: src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx [hunk 11]
```diff
@@ -322,29 +676,108 @@ cdef class DMDA(DM):
...
         cdef PetscInt ival = asInt(field)
         cdef const char *cval = NULL
         name = str2bytes(name, &cval)
         CHKERR( DMDASetFieldName(self.dm, ival, cval) )
 
-    def getFieldName(self, field):
+    def getFieldName(self, field: int) -> str:
+        """Return the name of an individual field component.
+
+        Not collective; name will contain a common value.
```
Comment (stefanozampini) [inline src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx:706]:
```suggestion
        Not collective.
```

Comment (stefanozampini) [reply src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx:706]: Not sure if you have other specifications of this type. If so, just remove them

File: src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx [hunk 12]
```diff
@@ -353,44 +786,168 @@ cdef class DMDA(DM):
...
+        Vec vg,
+        Vec vn,
+        addv: InsertMode | None = None,
+    ) -> None:
+        """Map values to the "natural" grid ordering.
+
+        Output Parameter
+        l - the natural ordering values
+        l - the natural ordering values
+        l - the global values in the natural ordering
```
Comment (stefanozampini) [inline src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx:861]: ?????

Comment (stefanozampini) [reply src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx:861]: ChatGPT.....

File: src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx [hunk 13]
```diff
@@ -353,44 +786,168 @@ cdef class DMDA(DM):
...
+
+        Notes
+        The global and natural vectors used here need not be the same as those obtained from DMCreateGlobalVector() and DMDACreateNaturalVector(), BUT they must have the same parallel data layout; they could, for example, be obtained with VecDuplicate() from the DMDA originating vectors.
+
+        You must call DMDACreateNaturalVector() before using this routine
+
+        Neighbor-wise collective.
```
Comment (stefanozampini) [inline src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx:868]: Collective or others go after the first line. Check in other places.

File: src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx [hunk 14]
```diff
@@ -353,44 +786,168 @@ cdef class DMDA(DM):
...
         CHKERR( DMDANaturalToGlobalBegin(self.dm, vn.vec, im, vg.vec) )
         CHKERR( DMDANaturalToGlobalEnd  (self.dm, vn.vec, im, vg.vec) )
 
     #
 
-    def getAO(self):
+    def getAO(self) -> AO:
+        """Return the application ordering context for a distributed array.
+
+        In this case, the AO maps to the natural grid ordering that would be
```
Comment (stefanozampini) [inline src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx:923]:
```suggestion
        The returned `AO` maps to the natural grid ordering that would be
```

File: src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx [hunk 15]
```diff
@@ -400,10 +957,30 @@ cdef class DMDA(DM):
 
     #
 
-    def setRefinementFactor(self,
-                            refine_x=2,
-                            refine_y=2,
-                            refine_z=2):
+    def setRefinementFactor(
+        self,
+        refine_x: int | None = 2,
```
Comment (stefanozampini) [inline src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx:962]: Why `| None` ? It is not the first time I found this, please double-check the rest of the file

File: src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx [hunk 16]
```diff
@@ -422,27 +1008,90 @@ cdef class DMDA(DM):
...
+        Not collective.
+
+        See Also
+        --------
+        setInterpolationType, petsc.DMDAGetInterpolationType
+
+        """
         cdef PetscDMDAInterpolationType ival = DMDA_INTERPOLATION_Q0
         CHKERR( DMDAGetInterpolationType(self.dm, &ival) )
         return <long>ival
```
Comment (stefanozampini) [inline src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx:1044]: Wow, this should be fixed. Ask @dalcinl

File: src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx [hunk 17]
```diff
@@ -422,27 +1008,90 @@ cdef class DMDA(DM):
...
+        See Also
+        --------
+        getElementType, petsc.DMDASetElementType
+
+        """
         cdef PetscDMDAElementType ival = daelementtype(elem_type)
         CHKERR( DMDASetElementType(self.dm, ival) )
 
-    def getElementType(self):
+    def getElementType(self) -> int:
```
Comment (stefanozampini) [inline src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx:1061]: Again, this should be fixed. Ask @dalcinl

File: src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx [hunk 18]
```diff
@@ -463,51 +1112,64 @@ cdef class DMDA(DM):
     #
 
     property dim:
-        def __get__(self):
+        """The dimension."""
```
Comment (stefanozampini) [inline src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx:1115]:
```suggestion
        """The grid dimension."""
```

File: src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx [hunk 19]
```diff
@@ -463,51 +1112,64 @@ cdef class DMDA(DM):
...
+        def __get__(self) -> int:
             return self.getDim()
 
     property dof:
-        def __get__(self):
+        """The number of dof associated with each stratum of the grid."""
```
Comment (stefanozampini) [inline src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx:1120]: dof or dofs? what did we decide? Must be diligent consistent monkeys

File: src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx [hunk 20]
```diff
@@ -463,51 +1112,64 @@ cdef class DMDA(DM):
...
 
     property sizes:
-        def __get__(self):
+        """The global dimension in each direction."""
+        def __get__(self) -> tuple[int, ...]:
             return self.getSizes()
 
     property proc_sizes:
-        def __get__(self):
+        """The number of ranks in each direction in the global decomposition."""
```
Comment (stefanozampini) [inline src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx:1130]: I don't like `ranks`, I prefer `processes`. Maybe something we should make uniform throughout the docs. @dalcinl ?

File: src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx [hunk 21]
```diff
@@ -463,51 +1112,64 @@ cdef class DMDA(DM):
...
+        def __get__(self) -> tuple[int, ...]:
             return self.getProcSizes()
 
+    # TODO: see the note above at getBoundaryType
```
Comment (stefanozampini) [inline src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx:1134]: What is this TODO? Typo?

File: src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx [hunk 22]
```diff
@@ -463,51 +1112,64 @@ cdef class DMDA(DM):
...
 
     property stencil:
-        def __get__(self):
+        """Stencil type and width."""
+        def __get__(self) -> tuple[StencilType, int]:
             return self.getStencil()
 
     property stencil_type:
-        def __get__(self):
+        """Elementwise ghost/halo stencil type."""
```
Comment (stefanozampini) [inline src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx:1146]:
```suggestion
        """Stencil type."""
```

File: src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx [hunk 23]
```diff
@@ -463,51 +1112,64 @@ cdef class DMDA(DM):
...
+        def __get__(self) -> str:
```
Comment (stefanozampini) [inline src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx:1147]: `str` or we have an enum?

**Review by stefanozampini — CHANGES_REQUESTED** [review]

**Comment by stefanozampini** [comment]
image

File: src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx [hunk 24]
```diff
@@ -311,7 +646,26 @@ cdef class DMDA(DM):
...
                 (toInt(m), toInt(n), toInt(p))[:<Py_ssize_t>dim])
 
-    def getGhostCorners(self):
+    def getGhostCorners(self) -> tuple[()] | tuple[tuple[int], tuple[int]] | tuple[tuple[int, int], tuple[int, int]] | tuple[tuple[int, int, int], tuple[int, int, int]]:
+        """Return lower left corner and size of local region in each dimension.
+
+        Returns the global (x,y,z) indices of the lower left corner (first
+        tuple) and size of the local region (second tuple).
+
+        Excluding ghost points.
```
Comment (mrogowski) [inline src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx:655]:
```suggestion
```

File: src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx [hunk 25]
```diff
@@ -311,7 +646,26 @@ cdef class DMDA(DM):
...
+
```
Comment (mrogowski) [inline src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx:656]:
```suggestion
```

File: src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx (hunk 12 above)
Comment (mrogowski) [inline src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx:861]:
```suggestion
```

File: src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx (hunk 16 above)
Comment (mrogowski) [reply src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx:1044]: He saw it and did not like it but managed to stopped himself from fixing it just yet. I'll add a FIXME.

File: src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx [hunk 26]
```diff
@@ -422,27 +1008,90 @@ cdef class DMDA(DM):
...
+        See Also
+        --------
+        getInterpolationType, petsc.DMDASetInterpolationType
+
+        """
         cdef PetscDMDAInterpolationType ival = dainterpolationtype(interp_type)
         CHKERR( DMDASetInterpolationType(self.dm, ival) )
 
-    def getInterpolationType(self):
+    def getInterpolationType(self) -> int:
```
Comment (mrogowski) [inline src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx:1032]:
```suggestion
    # FIXME: Return type
    def getInterpolationType(self) -> int:
```

File: src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx (hunk 17 above)
Comment (mrogowski) [inline src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx:1061]:
```suggestion
    # FIXME: Return type
    def getElementType(self) -> int:
```

File: src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx (hunk 19 above)
Comment (mrogowski) [inline src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx:1120]: DOFs 🐒

Comment (mrogowski) [inline src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx:1120]:
```suggestion
        """The number of DOFs associated with each stratum of the grid."""
```

File: src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx [hunk 27]
```diff
@@ -15,17 +15,76 @@ class DMDAElementType(object):
...
     InterpolationType = DMDAInterpolationType
     ElementType       = DMDAElementType
 
     #
 
-    def create(self, dim=None, dof=None,
-               sizes=None, proc_sizes=None, boundary_type=None,
-               stencil_type=None, stencil_width=None,
-               bint setup=True, ownership_ranges=None, comm=None):
+    # TODO: please verify
```
Comment (mrogowski) [inline src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx:26]:
```suggestion
```

File: src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx [hunk 28]
```diff
@@ -322,29 +676,108 @@ cdef class DMDA(DM):
...
 
     #
 
-    def setUniformCoordinates(self,
-                              xmin=0, xmax=1,
-                              ymin=0, ymax=1,
-                              zmin=0, zmax=1):
+    def setUniformCoordinates(
+        self,
+        xmin: float | None = 0,
```
[the remark this answers is not in the document]
Comment (mrogowski) [reply src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx:743]: I'll have to ask about this one tomorrow. I may have misunderstood something. 🐒

File: src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx (hunk 13 above)
Comment (mrogowski) [reply src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx:860]: That's another one for the mass search&replace. I got it this way everywhere.

File: src/binding/petsc4py/conf/cyautodoc.py [hunk 29]
```diff
@@ -1,6 +1,7 @@
 # ruff: noqa: UP008,UP031
 from Cython.Compiler import Options
 from Cython.Compiler import PyrexTypes
+from Cython.Compiler.ExprNodes import TupleNode
```
Comment (mrogowski) [inline src/binding/petsc4py/conf/cyautodoc.py:4]:
```suggestion
```

File: src/binding/petsc4py/conf/cyautodoc.py [hunk 30]
```diff
@@ -11,6 +12,18 @@
...
+        self.put(u"[")
+        if isinstance(node.index, TupleNode):
+            if node.index.subexpr_nodes():
+                self.emit_sequence(node.index)
+            else:
+                self.put(u"()")
+        else:
+            self.visit(node.index)
+        self.put(u"]")
+
```
Comment (mrogowski) [inline src/binding/petsc4py/conf/cyautodoc.py:26]:
```suggestion
```

File: src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx [hunk 31]
```diff
@@ -311,7 +645,24 @@ cdef class DMDA(DM):
...
+        """Return the lower left corner and the size of the local region in each dimension, including ghost points.
+
+        Returns the global (x,y,z) indices of the lower left corner (first
+        tuple) and size of the local region (second tuple).
+
+        The corner information is independent of the number of degrees of
+        freedom per node. Thus the returned values can be thought of as
+        coordinates on a logical grid, where each grid point has (potentially)
+        several degrees of freedom.
+
```
Comment (mrogowski) [inline src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx:658]:
```suggestion
```

File: src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx [hunk 32]
```diff
@@ -15,17 +15,75 @@ class DMDAElementType(object):
...
-               sizes=None, proc_sizes=None, boundary_type=None,
-               stencil_type=None, stencil_width=None,
-               bint setup=True, ownership_ranges=None, comm=None):
+    def create(
+        self,
+        dim: int | None = None,
+        dof: int | None = None,
+        sizes: tuple[int, ...] | None = None,
+        proc_sizes: tuple[int, ...] | None = None,
+        boundary_type: tuple[DM.BoundaryType, ...] | tuple[int, ...] | tuple[str, ...] | tuple[bool, ...] | None = None,
```
Comment (dalcinl) [inline src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx:32]: Same as in `DMStag`.

File: src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx [hunk 33]
```diff
@@ -82,8 +140,37 @@ cdef class DMDA(DM):
         PetscCLEAR(self.obj); self.dm = newda
         return self
 
-    def duplicate(self, dof=None, boundary_type=None,
-                  stencil_type=None, stencil_width=None):
+    def duplicate(
+        self,
+        dof: int | None = None,
+        boundary_type: tuple[DM.BoundaryType, ...] | tuple[int, ...] | tuple[str, ...] | tuple[bool, ...] | None = None,
```
Comment (dalcinl) [inline src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx:146]: Same as in `DMStag`.

File: src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx [hunk 34]
```diff
@@ -198,14 +377,41 @@ cdef class DMDA(DM):
                             NULL) )
         return toDims(dim, m, n, p)
 
-    def setBoundaryType(self, boundary_type):
+    def setBoundaryType(
+        self,
+        boundary_type: tuple[DM.BoundaryType, ...] | tuple[int, ...] | tuple[str, ...] | tuple[bool, ...],
```
Comment (dalcinl) [inline src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx:382]: Same as in `DMStag`.

File: src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx [hunk 35]
```diff
@@ -289,7 +592,19 @@ cdef class DMDA(DM):
                 (toInt(y), toInt(y+n)),
                 (toInt(z), toInt(z+p)))[:<Py_ssize_t>dim]
 
-    def getOwnershipRanges(self):
+    # TODO: this seems to be a sequence and not a tuple, right?
+    def getOwnershipRanges(self) -> tuple[Sequence[int], ...]:
```
Comment (dalcinl) [inline src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx:596]:
```suggestion
    def getOwnershipRanges(self) -> tuple[ArrayInt, ...]:
```

File: src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx [hunk 36]
```diff
@@ -322,29 +668,108 @@ cdef class DMDA(DM):
...
-                              ymin=0, ymax=1,
-                              zmin=0, zmax=1):
+    def setUniformCoordinates(
+        self,
+        xmin: float | None = 0,
+        xmax: float | None = 1,
+        ymin: float | None = 0,
+        ymax: float | None = 1,
+        zmin: float | None = 0,
+        zmax: float | None = 1,
```
Comment (dalcinl) [inline src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx:743]:
```suggestion
        xmin: float = 0,
        xmax: float = 1,
        ymin: float  = 0,
        ymax: float = 1,
        zmin: float = 0,
        zmax: float = 1,
```

File: src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx [hunk 37]
```diff
@@ -400,10 +941,30 @@ cdef class DMDA(DM):
...
 
-    def setRefinementFactor(self,
-                            refine_x=2,
-                            refine_y=2,
-                            refine_z=2):
+    def setRefinementFactor(
+        self,
+        refine_x: int | None = 2,
+        refine_y: int | None = 2,
+        refine_z: int | None = 2,
```
Comment (dalcinl) [inline src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx:948]:
```suggestion
        refine_x: int = 2,
        refine_y: int = 2,
        refine_z: int = 2,
```

File: src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx [hunk 38]
```diff
@@ -422,27 +992,92 @@ cdef class DMDA(DM):
...
+        --------
+        getInterpolationType, petsc.DMDASetInterpolationType
+
+        """
         cdef PetscDMDAInterpolationType ival = dainterpolationtype(interp_type)
         CHKERR( DMDASetInterpolationType(self.dm, ival) )
 
-    def getInterpolationType(self):
+    # FIXME: Return type
+    def getInterpolationType(self) -> int:
```
Comment (dalcinl) [inline src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx:1017]:
```suggestion
    def getInterpolationType(self) -> int:
```

File: src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx [hunk 39]
```diff
@@ -422,27 +992,92 @@ cdef class DMDA(DM):
...
+        --------
+        getElementType, petsc.DMDASetElementType
+
+        """
         cdef PetscDMDAElementType ival = daelementtype(elem_type)
         CHKERR( DMDASetElementType(self.dm, ival) )
 
-    def getElementType(self):
+    # FIXME: Return type
+    def getElementType(self) -> int:
```
Comment (dalcinl) [inline src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx:1047]:
```suggestion
    def getElementType(self) -> ElementType:
```

File: src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx (hunk 20 above)
Comment (dalcinl) [reply src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx:1116]: Go ahead.

File: src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx [hunk 40]
```diff
@@ -198,14 +377,41 @@ cdef class DMDA(DM):
...
+        """
         cdef PetscDMBoundaryType btx = DM_BOUNDARY_NONE
         cdef PetscDMBoundaryType bty = DM_BOUNDARY_NONE
         cdef PetscDMBoundaryType btz = DM_BOUNDARY_NONE
         asBoundary(boundary_type, &btx, &bty, &btz)
         CHKERR( DMDASetBoundaryType(self.dm, btx, bty, btz) )
 
-    def getBoundaryType(self):
+    # TODO: expected a string, but this seems to be an int
+    def getBoundaryType(self) -> tuple[int, ...]:
```
Comment (dalcinl) [inline src/binding/petsc4py/src/petsc4py/PETSc/DMDA.pyx:405]:
```suggestion
    def getBoundaryType(self) -> tuple[DM.BoundaryType, ...]:
```

**Comment by mrogowski** [comment]
@stefanozampini I'll reorder or "Collective" lines in a separate PR so you can approve once this is the only remaining complaint here.

**Review by stefanozampini — APPROVED** [review]

**Pull request merged**